lightning/src/routing/router.rs

```diff
@@ -1537,12 +1545,85 @@ where L::Target: Logger {
+fn add_random_cltv_offset(route: &mut Route, payment_params: &PaymentParameters, network: &NetworkGraph, random_seed_bytes: &[u8; 32]) {
+	let network_graph = network.read_only();
+	let network_channels = network_graph.channels();
+	let network_nodes = network_graph.nodes();
+
+	for path in route.paths.iter_mut() {
+		let mut shadow_ctlv_expiry_delta_offset: u32 = 0;
+
+		// Choose the last publicly known node as the starting point for the random walk
+		if let Some(starting_hop) = path.iter().rev().find(|h| network_nodes.contains_key(&NodeId::from_pubkey(&h.pubkey))) {
+			let mut cur_node_id = NodeId::from_pubkey(&starting_hop.pubkey);
+
+			// Init PRNG with path nonce
+			let mut path_nonce = [0u8; 12];
+			path_nonce.copy_from_slice(&cur_node_id.as_slice()[..12]);
+			let mut prng = ChaCha20::new(random_seed_bytes, &path_nonce);
+			let mut random_path_bytes = [0u8; ::core::mem::size_of::<usize>()];
+
+			// Pick a random path length in [1 .. 3]
+			prng.process_in_place(&mut random_path_bytes);
+			let random_walk_length = usize::from_be_bytes(random_path_bytes).wrapping_rem(3).wrapping_add(1);
+
+			for _random_hop in 0..random_walk_length {
+				if let Some(cur_node) = network_nodes.get(&cur_node_id) {
+					// Randomly choose the next hop
+					prng.process_in_place(&mut random_path_bytes);
+					if let Some(random_channel) = usize::from_be_bytes(random_path_bytes).checked_rem(cur_node.channels.len())
+						.and_then(|index| cur_node.channels.get(index))
+						.and_then(|id| network_channels.get(id)) {
+							random_channel.as_directed_to(&cur_node_id).map(|(dir_info, next_id)| {
+								dir_info.direction().map(|channel_update_info|
+									shadow_ctlv_expiry_delta_offset = shadow_ctlv_expiry_delta_offset
+										.checked_add(channel_update_info.cltv_expiry_delta.into())
+										.unwrap_or(shadow_ctlv_expiry_delta_offset));
+								cur_node_id = *next_id;
+							});
+						}
+				}
+			}
+		} else {
+			// If the entire path is private, choose a random offset from multiples of 40,
+			// which is the most prevalent cltv_expiry_delta currently used in the network
+			let mut prng = ChaCha20::new(random_seed_bytes, &[0u8; 8]);
+			let mut random_bytes = [0u8; 4];
+			prng.process_in_place(&mut random_bytes);
+			let random_walk_length = u32::from_be_bytes(random_bytes).wrapping_rem(3).wrapping_add(1);
+			shadow_ctlv_expiry_delta_offset = random_walk_length.wrapping_mul(40);
+		}
+
+		// Limit the total offset to reduce the worst-case locked liquidity timevalue
+		const MAX_SHADOW_CLTV_EXPIRY_DELTA_OFFSET: u32 = 3*144;
+		shadow_ctlv_expiry_delta_offset = cmp::min(shadow_ctlv_expiry_delta_offset, MAX_SHADOW_CLTV_EXPIRY_DELTA_OFFSET);
+
+		// Limit the offset so we never exceed the max_total_cltv_expiry_delta
+		let max_path_offset = payment_params.max_total_cltv_expiry_delta
+			.checked_sub(path.iter().map(|h| h.cltv_expiry_delta).sum())
+			.unwrap_or(shadow_ctlv_expiry_delta_offset);
```

Comment:
Shouldn't this be `unwrap_or(0)`? If our path expiry is more than the max already we probably shouldn't add an offset at all? Also, were we planning on limiting the total CLTV during the `get_route` run to give us a little headroom here?

Reply:
Well my thinking here was that since we limit the route overall with `max_total_cltv_expiry_delta`, this subtraction should never fail. If however something unexpected would occur and it would fail, we might be better off to default to provide some privacy (capped at `3*144`), instead of none. But it can of course also default to `0`, I have no strong preference here.

As for the considering the offset in `get_route`: [after looking at the data](https://github.com/lightningdevkit/rust-lightning/pull/1286#discussion_r809896779) I found that there were basically no cases in which there was not ample room to add the offset and hence thought it would be cleaner not to interfere with pathfinding and keep the two methods entirely separate. However, if we still want to take some of the offset into account, I can of course go back to, for example, subtracting `2*40` or `2*144` from the limit during path finding.

Reply:
> Well my thinking here was that since we limit the route overall with max_total_cltv_expiry_delta, this subtraction should never fail.

Right, sorry, indeed.

>  If however something unexpected were to happen and it would fail, we might be better off to default to provide some privacy (capped at 3*144), instead of none.

:man_shrugging: I guess we shouldn't ever violate the user-provided value ever?

> As for the considering the offset in get_route: https://github.com/lightningdevkit/rust-lightning/pull/1286#discussion_r809896779 I found that there were basically no cases in which there was not ample room to add the offset and hence thought it would be cleaner not to interfere with pathfinding and keep the two methods entirely separate.

Hmm, I feel like that's a better reason to interfere than to not interfere.

I'm still a bit confused by this hunk here - why do we do the subtraction at `max_path_offset.checked_sub(max_path_offset.wrapping_rem(40))`?

Reply:
> 🤷‍♂️ I guess we shouldn't ever violate the user-provided value ever?
> ...
> Hmm, I feel like that's a better reason to interfere than to not interfere.
Alright, both addressed by https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/f7be9daded106710d969ddfcc6c3a949a05b906c.

> I'm still a bit confused by this hunk here - why do we do the subtraction at `max_path_offset.checked_sub(max_path_offset.wrapping_rem(40))`?
The difference of the path total to `max_total_cltv_expiry_delta` may be a pretty uncommon value, which could result in an implausible shadow route offset. The thinking here therefore was to choose the path offset to be the closest multiple of 40 to increase plausibility. I now had another look at this part and added a comment to make this more transparent in https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/f7be9daded106710d969ddfcc6c3a949a05b906c.